<h1 align = "center"><font>BEST MOVIE ANALYZER. ROAD TO THE MODEL</font></h1>

In [2]:
#All libs we are required (while project is not finished the list will become more biger!):
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

<h2 align = "center"><font>FIRST LOOK AT DATA</font></h2>
<h3 align = "center"><font color='#4e78f5'>Our dataset in general:</font></h3>

In [3]:
#First look at head of dataset:
movie_df_all = pd.read_csv('movieset_6800.csv')
movie_df_all.shape

(7668, 15)

In [4]:
movie_df_all.head(5)

name rating      genre  year  \
0                                     The Shining      R      Drama  1980   
1                                 The Blue Lagoon      R  Adventure  1980   
2  Star Wars: Episode V - The Empire Strikes Back     PG     Action  1980   
3                                       Airplane!     PG     Comedy  1980   
4                                      Caddyshack      R     Comedy  1980   

                        released  score      votes         director  \
0  June 13, 1980 (United States)    8.4   927000.0  Stanley Kubrick   
1   July 2, 1980 (United States)    5.8    65000.0   Randal Kleiser   
2  June 20, 1980 (United States)    8.7  1200000.0   Irvin Kershner   
3   July 2, 1980 (United States)    7.7   221000.0     Jim Abrahams   
4  July 25, 1980 (United States)    7.3   108000.0     Harold Ramis   

                    writer            star         country      budget  \
0             Stephen King  Jack Nicholson  United Kingdom  19000000.0   
1  Henry De Vere Stacpoole  Brooke Shields   United States   4500000.0   
2           Leigh Brackett     Mark Hamill   United States  18000000.0   
3             Jim Abrahams     Robert Hays   United States   3500000.0   
4       Brian Doyle-Murray     Chevy Chase   United States   6000000.0   

         gross             company  runtime  
0   46998772.0        Warner Bros.    146.0  
1   58853106.0   Columbia Pictures    104.0  
2  538375067.0           Lucasfilm    124.0  
3   83453539.0  Paramount Pictures     88.0  
4   39846344.0      Orion Pictures     98.0

<h3 align = "center"><font color='#4e78f5'>General situation:</font></h3>

<p>We have 7668 movies and different inforamtion about this movies.</p>
<p>Name - name of the movie.</p>
<p>Rating - who can watch this movie.</p>
<p>Genre - groups of feature films, distinguished on the basis of similar features of their internal structure.</p>
<p>Year - released year.</p>
<p>Released - released date with additional information.</p>
<p>Score - IMDB score.</p>
<p>Votes - number of votes.</p>
<p>Director - director who was produced film.</p>
<p>Writer - person who wrote history.</p>
<p>Star - the main actor of the movie.</p>
<p>Budget - quntity of money spent to the film.</p>
<p>Gross - quantity of money movie earned.</p>
<p>Company - company that produced the movie.</p>
<p>Runtime - duration of the movie</p>

<h2 align = "center"><font>CLEANING FEATURE</font></h2>

In [5]:
movie_df_all = movie_df_all.dropna(axis='index', how='any', subset=['score'])
movie_df_all

name rating      genre  year  \
0                                        The Shining      R      Drama  1980   
1                                    The Blue Lagoon      R  Adventure  1980   
2     Star Wars: Episode V - The Empire Strikes Back     PG     Action  1980   
3                                          Airplane!     PG     Comedy  1980   
4                                         Caddyshack      R     Comedy  1980   
...                                              ...    ...        ...   ...   
7660                                Aloha Surf Hotel    NaN     Comedy  2020   
7663                                    More to Life    NaN      Drama  2020   
7664                                     Dream Round    NaN     Comedy  2020   
7665                                   Saving Mbango    NaN      Drama  2020   
7667                                       Tee em el    NaN     Horror  2020   

                              released  score      votes            director  \
0        June 13, 1980 (United States)    8.4   927000.0     Stanley Kubrick   
1         July 2, 1980 (United States)    5.8    65000.0      Randal Kleiser   
2        June 20, 1980 (United States)    8.7  1200000.0      Irvin Kershner   
3         July 2, 1980 (United States)    7.7   221000.0        Jim Abrahams   
4        July 25, 1980 (United States)    7.3   108000.0        Harold Ramis   
...                                ...    ...        ...                 ...   
7660  November 5, 2020 (United States)    7.1       14.0  Stefan C. Schaefer   
7663  October 23, 2020 (United States)    3.1       18.0       Joseph Ebanks   
7664  February 7, 2020 (United States)    4.7       36.0        Dusty Dukatz   
7665         April 27, 2020 (Cameroon)    5.7       29.0        Nkanya Nkwai   
7667   August 19, 2020 (United States)    5.7        7.0        Pereko Mosia   

                       writer              star         country      budget  \
0                Stephen King    Jack Nicholson  United Kingdom  19000000.0   
1     Henry De Vere Stacpoole    Brooke Shields   United States   4500000.0   
2              Leigh Brackett       Mark Hamill   United States  18000000.0   
3                Jim Abrahams       Robert Hays   United States   3500000.0   
4          Brian Doyle-Murray       Chevy Chase   United States   6000000.0   
...                       ...               ...             ...         ...   
7660       Stefan C. Schaefer       Augie Tulba   United States         NaN   
7663            Joseph Ebanks      Shannon Bond   United States      7000.0   
7664              Lisa Huston  Michael Saquella   United States         NaN   
7665             Lynno Lovert      Onyama Laura   United States     58750.0   
7667             Pereko Mosia  Siyabonga Mabaso    South Africa         NaN   

            gross                    company  runtime  
0      46998772.0               Warner Bros.    146.0  
1      58853106.0          Columbia Pictures    104.0  
2     538375067.0                  Lucasfilm    124.0  
3      83453539.0         Paramount Pictures     88.0  
4      39846344.0             Orion Pictures     98.0  
...           ...                        ...      ...  
7660          NaN        Abominable Pictures     90.0  
7663          NaN                        NaN     90.0  
7664          NaN  Cactus Blue Entertainment     90.0  
7665          NaN           Embi Productions      NaN  
7667          NaN                PK 65 Films    102.0  

[7665 rows x 15 columns]

<h2 align = "center"><font>RATE FEATURE</font></h2>
<h3 align = "center"><font color='#6ad156'>One-hot-encoding rate feature:</font></h3>

In [6]:
#Sepparte columns with rate
movie_df_ohe = pd.get_dummies(movie_df_all, columns = ['rating'])
movie_df_ohe

name      genre  year  \
0                                        The Shining      Drama  1980   
1                                    The Blue Lagoon  Adventure  1980   
2     Star Wars: Episode V - The Empire Strikes Back     Action  1980   
3                                          Airplane!     Comedy  1980   
4                                         Caddyshack     Comedy  1980   
...                                              ...        ...   ...   
7660                                Aloha Surf Hotel     Comedy  2020   
7663                                    More to Life      Drama  2020   
7664                                     Dream Round     Comedy  2020   
7665                                   Saving Mbango      Drama  2020   
7667                                       Tee em el     Horror  2020   

                              released  score      votes            director  \
0        June 13, 1980 (United States)    8.4   927000.0     Stanley Kubrick   
1         July 2, 1980 (United States)    5.8    65000.0      Randal Kleiser   
2        June 20, 1980 (United States)    8.7  1200000.0      Irvin Kershner   
3         July 2, 1980 (United States)    7.7   221000.0        Jim Abrahams   
4        July 25, 1980 (United States)    7.3   108000.0        Harold Ramis   
...                                ...    ...        ...                 ...   
7660  November 5, 2020 (United States)    7.1       14.0  Stefan C. Schaefer   
7663  October 23, 2020 (United States)    3.1       18.0       Joseph Ebanks   
7664  February 7, 2020 (United States)    4.7       36.0        Dusty Dukatz   
7665         April 27, 2020 (Cameroon)    5.7       29.0        Nkanya Nkwai   
7667   August 19, 2020 (United States)    5.7        7.0        Pereko Mosia   

                       writer              star         country  ...  \
0                Stephen King    Jack Nicholson  United Kingdom  ...   
1     Henry De Vere Stacpoole    Brooke Shields   United States  ...   
2              Leigh Brackett       Mark Hamill   United States  ...   
3                Jim Abrahams       Robert Hays   United States  ...   
4          Brian Doyle-Murray       Chevy Chase   United States  ...   
...                       ...               ...             ...  ...   
7660       Stefan C. Schaefer       Augie Tulba   United States  ...   
7663            Joseph Ebanks      Shannon Bond   United States  ...   
7664              Lisa Huston  Michael Saquella   United States  ...   
7665             Lynno Lovert      Onyama Laura   United States  ...   
7667             Pereko Mosia  Siyabonga Mabaso    South Africa  ...   

      rating_NC-17  rating_Not Rated rating_PG  rating_PG-13  rating_R  \
0                0                 0         0             0         1   
1                0                 0         0             0         1   
2                0                 0         1             0         0   
3                0                 0         1             0         0   
4                0                 0         0             0         1   
...            ...               ...       ...           ...       ...   
7660             0                 0         0             0         0   
7663             0                 0         0             0         0   
7664             0                 0         0             0         0   
7665             0                 0         0             0         0   
7667             0                 0         0             0         0   

      rating_TV-14  rating_TV-MA  rating_TV-PG  rating_Unrated  rating_X  
0                0             0             0               0         0  
1                0             0             0               0         0  
2                0             0             0               0         0  
3                0             0             0               0         0  
4                0             0             0               0         0  
...   

In [7]:
#Create my rate system where M_A - middle audience. U_A - universal audience. A_A - adult audeince

#M_A
movie_df_ohe["Rate_M_A"] = (movie_df_ohe["rating_PG"] + movie_df_ohe["rating_PG-13"] + 
movie_df_ohe["rating_TV-PG"])

#U_A
movie_df_ohe["Rate_U_A"] = (movie_df_ohe["rating_G"] + movie_df_ohe["rating_TV-14"] + movie_df_ohe["rating_Approved"])

#A_A
movie_df_ohe["Rate_A_A"] = (movie_df_ohe["rating_NC-17"] + movie_df_ohe["rating_Not Rated"] + movie_df_ohe["rating_R"]
+ movie_df_ohe["rating_TV-MA"] + movie_df_ohe["rating_Unrated"] + movie_df_ohe["rating_X"])

#Delete useless columns
movie_df_ohe = movie_df_ohe.drop(columns = ["rating_PG", "rating_PG-13", "rating_TV-PG",
 "rating_TV-14", "rating_NC-17", "rating_Not Rated", "rating_R", "rating_TV-MA", "rating_Unrated", "rating_X", "rating_G", "rating_Approved"])

movie_df_ohe.shape

(7665, 17)

<h3 align = "center"><font color='#6ad156'>Rate feature correct for model learning:</font></h3>

In [8]:
#Current DataSet:
movie_df_ohe[["Rate_A_A", "Rate_M_A", "Rate_U_A"]]

Rate_A_A  Rate_M_A  Rate_U_A
0            1         0         0
1            1         0         0
2            0         1         0
3            0         1         0
4            1         0         0
...        ...       ...       ...
7660         0         0         0
7663         0         0         0
7664         0         0         0
7665         0         0         0
7667         0         0         0

[7665 rows x 3 columns]

In [9]:
#General information about or new columns:
print("U_A count:", movie_df_ohe['Rate_U_A'].sum())

print("A_A count:", movie_df_ohe['Rate_A_A'].sum())

print("M_A count:", movie_df_ohe['Rate_M_A'].sum())

print("Nan count:", len(movie_df_ohe) - 
(movie_df_ohe['Rate_U_A'].sum() + movie_df_ohe['Rate_A_A'].sum() + movie_df_ohe['Rate_M_A'].sum()))

U_A count: 155
A_A count: 4066
M_A count: 3369
Nan count: 75


<h2 align = "center"><font>GENRE FEATURE</font></h2>
<h3 align = "center"><font color='#2db3c2'>One-hot-encoding genre feature</font></h3>

In [10]:
#General data analyzes
movie_df_ohe.value_counts("genre")

genre
Comedy       2245
Action       1704
Drama        1516
Crime         551
Biography     443
Adventure     427
Animation     338
Horror        322
Fantasy        44
Mystery        20
Thriller       16
Family         11
Romance        10
Sci-Fi         10
Western         3
Musical         2
Sport           1
Music           1
History         1
dtype: int64

In [11]:
#Sepparte columns with genre due to the hot-encoding technique
movie_df_ohe = pd.get_dummies(movie_df_ohe, columns = ['genre'])
movie_df_ohe.shape

(7665, 35)

<h2 align = "center"><font>STAR FEATURE</font></h2>
<h3 align = "center"><font color='#2db3c2'>Can we use it?</font></h3>

In [12]:
movie_df_ohe.value_counts("star")

star
Nicolas Cage         43
Robert De Niro       41
Tom Hanks            41
Denzel Washington    37
Bruce Willis         34
                     ..
Jenjira Pongpas       1
Jeni Courtney         1
Jena Malone           1
Jemma Dallender       1
Émilie Dequenne       1
Length: 2811, dtype: int64

<h2 align = "center"><font>FIRST MODEL</font></h2>
<h3 align = "center"><font color='#4e78f5'>Delete all columns wich are nominal</font></h3>

In [13]:
#Delete and check what we have
move_model_set = movie_df_ohe.drop(columns = ["name", "released", "director", "star", "writer", "country", "company"]).dropna()
move_model_set.dtypes

year                 int64
score              float64
votes              float64
budget             float64
gross              float64
runtime            float64
Rate_M_A             uint8
Rate_U_A             uint8
Rate_A_A             uint8
genre_Action         uint8
genre_Adventure      uint8
genre_Animation      uint8
genre_Biography      uint8
genre_Comedy         uint8
genre_Crime          uint8
genre_Drama          uint8
genre_Family         uint8
genre_Fantasy        uint8
genre_History        uint8
genre_Horror         uint8
genre_Music          uint8
genre_Musical        uint8
genre_Mystery        uint8
genre_Romance        uint8
genre_Sci-Fi         uint8
genre_Sport          uint8
genre_Thriller       uint8
genre_Western        uint8
dtype: object

<h3 align = "center"><font color='#4e78f5'>Lets train our model!</font></h3>

In [14]:
#Create DataSets for traning and testing (75, 25):
y = move_model_set.drop(move_model_set.columns.difference(["score"]), axis=1)
X = move_model_set.drop(columns = ["score"])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

rf = RandomForestRegressor(n_estimators=100, max_depth=4, random_state=42).fit(X_train, np.ravel(y_train))

In [23]:
move_model_set.columns

Index(['year', 'score', 'votes', 'budget', 'gross', 'runtime', 'Rate_M_A',
       'Rate_U_A', 'Rate_A_A', 'genre_Action', 'genre_Adventure',
       'genre_Animation', 'genre_Biography', 'genre_Comedy', 'genre_Crime',
       'genre_Drama', 'genre_Family', 'genre_Fantasy', 'genre_History',
       'genre_Horror', 'genre_Music', 'genre_Musical', 'genre_Mystery',
       'genre_Romance', 'genre_Sci-Fi', 'genre_Sport', 'genre_Thriller',
       'genre_Western'],
      dtype='object')

In [15]:
mean_squared_error(y_test['score'] , rf.predict(X_test), squared=False)

0.6998186340419529

<h3 align = "center"><font color='#4e78f5'>Function to rate your movie!</font></h3>

In [36]:
#Fucntion for your own movie parameters!
def try_own_movie(_year, _votes, _budget, _gross, _runtime, _rate, _genre):

    rate_df = []

    list_genre_to_add = []

    temp_genre = {"Action": 0, 
    "Adventure": 0, 
    "Animation": 0, 
    'Biography': 0, 
    'Comedy': 0, 
    'Crime': 0, 
    'Drama': 0,
    'Family': 0, 
    'Fantasy': 0,
    'History': 0, 
    'Horror': 0, 
    'Music': 0, 
    'Musical': 0, 
    'Mystery': 0, 
    'Romance': 0, 
    'Sci-Fi': 0,  
    'Sport': 0, 
    'Thriller': 0, 
    'Western': 0}

    for key in temp_genre:
        if key == _genre:
            list_genre_to_add.append(1)
        else:
            list_genre_to_add.append(0)

    if _rate == 'A':
        rate_df = [0, 0, 1]

    elif _rate == 'U':
        rate_df = [0, 1, 0]

    elif _rate == 'M':
        rate_df = [1, 0, 0] 

    temp_df = [_year, _votes, _budget, _gross, _runtime] + rate_df

    temp_df = temp_df + list_genre_to_add

    print("Input data:", temp_df, "Rate:", rf.predict([temp_df]))
    

<h3 align = "center"><font color='#4e78f5'>Try it by yourself!</font></h3>

In [38]:
#input data:
year = ... #year of movie
votes = ... #amount of votes
budget = ... #money spent for movie
gross = ... #budget of movie
runtime = ... #lenght of movie
rate = ... #rate A (adult), U (universal), M (Check attached doc)
genre = ... #select genre (Action, Drama, Crime etc.)
try_own_movie(year, votes, budget, gross, runtime, rate, genre)